### PR TITLE
Adds Login page

### PR DESCRIPTION
**Pull request by BrunoMartins11**
* Adds login page;
* Fetches safira token if credentials valid;
* Shows pop up if credentials invalid;
* Token is not being saved in the device for a future login or requests(will be worked on in the future);
* Introduces modularity into the app.

### CODE REVIEW

**Review by herulume — DISMISSED** [review]
lgtm
Good job!

File: lib/services/authentication.dart [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  Authentication({this.token, this.valid});
+
+  factory Authentication.fromJson(Map<String, dynamic> json) {
+    return Authentication(token: json['jwt'], valid: true);
+  }
+}
+
+Future<Authentication> fetchAuthToken(String email, String password) async {
+  
+  final response = await http.post('https://safira20.herokuapp.com/api/auth/sign_in?email=' + email + '&password=' + password);
```
Comment (apedrob) [inline lib/services/authentication.dart]: The api url should be on .env file

File: test/widget_test.dart [hunk 2]
```diff
@@ -4,27 +4,28 @@
...
-  });
-}
+//
+//import 'package:flutter/material.dart';
+//import 'package:flutter_test/flutter_test.dart';
+//
+//import 'package:nativemoon/main.dart';
+//import 'package:nativemoon/routes.dart';
+//
+//void main() {
```
Comment (apedrob) [inline test/widget_test.dart]: You should remove commented code. Why is this commented?

File: lib/components/myTextField.dart [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    this.filled,
+    this.filledColor,
+    this.textColor,
+    this.border,
+    this.hideText
+  );
+
+@override
+  Widget build(BuildContext context){
+    if(hideText){
```
Comment (apedrob) [inline lib/components/myTextField.dart]: Is this `if` right? You should use a [ternary operator](https://code-maven.com/slides/dart-programming/ternary-operator) for `labelStyle`

**Review by apedrob — CHANGES_REQUESTED** [review]

File: lib/services/authentication.dart (hunk 1 above)
Comment (herulume) [reply lib/services/authentication.dart]: Didn't comment anything since it isn't the aim of this PR. I do agree it could be done here.
Up to you guys.

File: test/widget_test.dart (hunk 2 above)
Comment (BrunoMartins11) [reply test/widget_test.dart]: This file is for testing purposes created in the beginning by flutter itself. When I changed the main only to create a Routes instance it started to throw errors. I decided to comment it for now.

File: lib/components/myTextField.dart (hunk 3 above)
Comment (BrunoMartins11) [reply lib/components/myTextField.dart]: Nice catch. I didn't know I could do it. I'll update this PR.

File: lib/services/authentication.dart (hunk 1 above)
Comment (ZePedroResende) [reply lib/services/authentication.dart]: The api url should be in a env file or even on a Service object. 
It shouldn't be a hardcoded or even on multiple points in the codebase because a simple domain change and you need to do a hole commit changing it. It should only depend on the env in what the app is build on.
The route should stay though.
Something like

`url[ENV] + '/api/auth/sign_in?email=' + email + '&password=' + password`

File: test/widget_test.dart (hunk 2 above)
Comment (ZePedroResende) [reply test/widget_test.dart]: Please try to add a test for this component. If not don't leave a file full of comment with no use.

File: lib/services/authentication.dart (hunk 1 above)
Comment (herulume) [reply lib/services/authentication.dart]: My point wasn't that it isn't needed. Only that he might prefer to do it in the next PR/a PR just for it. 

Why should the route stay?
I'd put `/api/auth/sign_in` in the env file as well.

Comment (BrunoMartins11) [reply lib/services/authentication.dart]: I agree with you @apedrob, it makes the app better structured.I will create one and update this PR.

Comment (ZePedroResende) [reply lib/services/authentication.dart]: I think the route should stay because its not duplicated anywhere else, and its not interchangeable by env. You are defining a middleware for the service, domains change, api dont unless its a breaking change and that should be changed all around. @herulume.
Also i think this is the correct time to address this small problems, because a good project setup saves a lot of time in the future with a possible migration in a large codebase.

File: test/widget_test.dart (hunk 2 above)
Comment (BrunoMartins11) [reply test/widget_test.dart]: Fixed.

File: lib/services/authentication.dart (hunk 1 above)
Comment (herulume) [reply lib/services/authentication.dart]: My train of thought was since it is API specific-ish (the sign in is pretty generic though) it should go the the env file.

Comment (BrunoMartins11) [reply lib/services/authentication.dart]: @herulume @apedrob @ZePedroResende  what do you think of this approach to environments I've just commited?

File: lib/services/authentication.dart [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  Authentication({this.token, this.valid});
+
+  factory Authentication.fromJson(Map<String, dynamic> json) {
+    return Authentication(token: json['jwt'], valid: true);
+  }
+}
+
+Future<Authentication> fetchAuthToken(String email, String password) async {
+  
+  final response = await http.post(env.baseUrl + '/api/auth/sign_in?email=' + email + '&password=' + password);
```
Comment (BrunoMartins11) [inline lib/services/authentication.dart]: @apedrob @ZePedroResende @herulume what do you think? I've redone the environment side of the app.

Comment (apedrob) [reply lib/services/authentication.dart]: The env variables shouldn't be on the source code or at least version control. Normally we use `dotenv` in JS, the same exists in [dart](https://pub.dev/packages/dotenv). Maybe @herulume knows better about this

Comment (BrunoMartins11) [reply lib/services/authentication.dart]: Those are external libraries implemented by the community, I tried not to create dependencies. The only way I found was this one.

Comment (BrunoMartins11) [reply lib/services/authentication.dart]: But if it's OK to use them I can change the implementation.

Comment (ZePedroResende) [reply lib/services/authentication.dart]: I think the thing that should be underlined here is that this file with urls and other env vars cant be tracked, just a sample of it.

Comment (herulume) [reply lib/services/authentication.dart]: I'm not sure if it's really used in dart but if `dotenv` is available I think it fits our needs perfectly @apedrob. We shouldn't do that work manually, fully agree with you. 
I would add a `.env.sample` file and tracked it as @ZePedroResende said.

TL;DR for me:
- Remove the `env.dart`
- Add `dotenv`
- Add a `.env.sample` file

Just a small note: there are some places in the code that still need to be replaced.

Comment (BrunoMartins11) [reply lib/services/authentication.dart]: I ended up using ``` flutter_dotenv  ``` not  ```dotenv ```, I found it more specific for flutter.

**Comment by herulume** [comment]
I would put the `.env.sample` in the nativemoon folder instead, but let us see what the others have to say.

File: lib/routes.dart [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+class Routes {
+
+  var routes = <String, WidgetBuilder>{
+    "/Home": (BuildContext context) => new HomePage()
+  };
+
+  Routes() {
+    runApp(new MaterialApp(
+      title: "Flutter Flat App",
```
Comment (herulume) [inline lib/routes.dart]: Not sure if this title is relevant at all, but maybe something more akin to nativemoon? Once again, not sure if relevant.

Comment (BrunoMartins11) [reply lib/routes.dart]: It's not relevant for the build of the app but makes sense to change it to nativemoon.

**Comment by nelsonmestevao** [comment]
> I would put the `.env.sample` in the nativemoon folder instead, but let us see what the others have to say.

I agree with @herulume :point_up:

File: lib/screens/Home/index.dart [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+class HomePage extends StatefulWidget {
+const HomePage({ Key key }) : super(key: key);
+
+@override
+HomePageState createState() => new HomePageState();
+
+}
+
+class HomePageState extends State<HomePage>{
```
Comment (herulume) [inline lib/screens/Home/index.dart:11]: Shouldn't this be private?

**Comment by BrunoMartins11** [comment]
> > I would put the `.env.sample` in the nativemoon folder instead, but let us see what the others have to say.
> 
> I agree with @herulume

Done

File: lib/screens/Login/index.dart [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+      if (auth.valid) {
+        //needs to save token somewhere
+        Navigator.pushNamed(context, "/Home");
+      } else {
+        PopUpAlert.showAlert(
+            context, "Error", "Invalid Email or Password", "OK");
+      }
+    }
+
+    final loginButon = new RoundedButton("Login", Colors.orange[200], 5.0, 0, 0, 30.0, _auth);
```
Comment (herulume) [inline lib/screens/Login/index.dart]: Button

**Review by herulume — CHANGES_REQUESTED** [review]

File: lib/screens/Home/index.dart (hunk 6 above)
Comment (BrunoMartins11) [reply lib/screens/Home/index.dart:11]: In my opinion no. In the docs and when you create a new base project the standard is this way.

**Review by herulume — APPROVED** [review]

**Review by apedrob — APPROVED** [review]

**Review by apedrob — APPROVED** [review]
🤖
